giraph-block-app/src/main/java/org/apache/giraph/block_app/library/Pieces.java

```diff
@@ -320,6 +325,91 @@ public String toString() {
   }
 
   /**
+   * Like reduceAndBroadcast, but uses array of handles for reducers and
+   * broadcasts, to make it feasible and performant when values are large.
+   * Each supplied value to reduce will be reduced in the handle defined by
+   * handleHashSupplier%numHandles
+   *
+   * @param <S> Single value type, objects passed on workers
+   * @param <R> Reduced value type
+   * @param <I> Vertex id type
+   * @param <V> Vertex value type
+   * @param <E> Edge value type
+   */
+  public static
+  <S, R extends Writable, I extends WritableComparable, V extends Writable,
+      E extends Writable>
+  Piece<I, V, E, NoMessage, Object> reduceAndBroadcastWithArrayOfHandles(
+      final String name,
+      final int numHandles,
+      final ReduceOperation<S, R> reduceOp,
+      final SupplierFromVertex<I, V, E, Long> handleHashSupplier,
+      final SupplierFromVertex<I, V, E, S> valueSupplier,
+      final ConsumerWithVertex<I, V, E, R> reducedValueConsumer) {
+    return new Piece<I, V, E, NoMessage, Object>() {
+      protected ArrayOfHandles.ArrayOfReducers<S, R> reducers;
+      protected BroadcastArrayHandle<R> broadcasts;
+
+      private int getHandleIndex(Vertex<I, V, E> vertex) {
+        return (int) Math.abs(handleHashSupplier.get(vertex) % numHandles);
+      }
+
+      @Override
+      public void registerReducers(
+          final CreateReducersApi reduceApi, Object executionStage) {
+        reducers = new ArrayOfHandles.ArrayOfReducers<>(
+            numHandles,
+            new Supplier<ReducerHandle<S, R>>() {
+              @Override
+              public ReducerHandle<S, R> get() {
+                return reduceApi.createLocalReducer(reduceOp);
```

Comment:
Actually, because you're using an ArrayOfHandles, this is going to be serialized-deserialized as a whole. This should be ok.

Though, this also assumes that reduce operations are stateless. In general, they are but there's nothing that prevents from writing a stateful one. Maybe add a comment about this.